### PR TITLE
mac80211: fix legacy rate decode

### PR DESCRIPTION
**Pull request by lxp**
Fix a null pointer dereference in legacy rate decoding. See: https://github.com/openwrt/openwrt/issues/13198

### CODE REVIEW

**Comment by xback** [comment]
Any idea why it could be null? :)

**Comment by lxp** [comment]
> Any idea why it could be null? :)

I already mentioned some theories in the last paragraph here: https://github.com/openwrt/openwrt/issues/13198#issuecomment-1800636494
However, I am not that familiar with the mac80211 code that I really can judge, what is most likely the case.
This should be a hotfix to avoid unnecessary kernel panics, when the problem occurs.
People like me are already plagued by these regular crashes since a few months, so I opt for fixing this in the stable version as soon as possible.

**Comment by lxp** [comment]
I hope you are lucky to reproduce it, as it seems to be very hard to trigger if you are not already suffering from the problem.
I think the following conditions must be met, so that there is a chance to reproduce it:
* You must have 2 GHz and 5 GHz radios enabled
* Clients must roam between 2 GHz and 5 GHz
* It may be necessary to have more than one ESSID on both radios and have the same ESSIDs on both radios
* It may happen less often if the wireless status is queried periodically e.g. by collectd
* It may happen less often if legacy rates are enabled on the 2.4 GHz radio
* It might happen only with specific clients (People mention Apple devices, but I think for me the problem was also triggered by an HP laptop with Intel Wi-Fi card running Linux)

**Comment by raenye** [comment]
Why only for 23.05?

File: package/kernel/mac80211/patches/subsys/341-mac80211-fix-legacy-rate-decode.patch
Comment (dangowrt) [inline package/kernel/mac80211/patches/subsys/341-mac80211-fix-legacy-rate-decode.patch:1]: Please refresh this patch using
```
make package/mac80211/{clean,refresh}
```

**Comment by lxp** [comment]
> Why only for 23.05?

I only have one production device with 23.05 running into this problem, so I could only test the fix in the 23.05 branch.
However, according to reports of other users, it also seems to be a problem in snapshot, so it might be worth also applying the fix there.

File: package/kernel/mac80211/patches/subsys/341-mac80211-fix-legacy-rate-decode.patch
Comment (lxp) [reply package/kernel/mac80211/patches/subsys/341-mac80211-fix-legacy-rate-decode.patch]: Thanks for your review and instructions on how to refresh the patch.
I squashed the changes and pushed the new version.

**Comment by trunneml** [comment]
Is there a "simple way" to build the ipk of this kmod and install it on my banana pi r3 running 23.5.2 stable?

**Comment by blocktrron** [comment]
I'll pick #14162 to 23.05 which accomplishes the same goal by providing some information what is going wrong.

**Comment by lxp** [comment]
> I'll pick #14162 to 23.05 which accomplishes the same goal by providing some information what is going wrong.

This is fine for me. We can close this pull request, once your PR is merged into 23.05. I am already looking forward to having this crash fixed in a 23.05 stable release :)

**Review by pspeder — APPROVED** [review]

**Comment by blocktrron** [comment]
Closed in favor of #14162

**Comment by roelofz** [comment]
Hi,

I am facing this issue on my BPI-R3, it is almost worth a brick, crashing every 24 hours at random. How do I know if this patch is applied.
Being very desperate I installed the latest snapshot, hoping this is included.
Is it possible to install the patch as a ... patch or do I need to wait for a new release, or is it all ready included in the latest snapshot? I hope it was not included in the latest stable branch as this is also giving me this issue.

Thanks for helping me out.

**Comment by roelofz** [comment]
How can I see if this fix is implemented in the stable (or another) release?

**Comment by essys** [comment]
I have the same question, I stopped using my BpiR3 due to this bug. It was basically unusable as it locked up at least once each day.
How can we check or push for a fix?
Will it be released in the next stable?

**Comment by winstonrc** [comment]
I arrived here from #12143. Since #14083 and #14162 both appear to be closed, I wanted to add a trail of where the issue stands.

It appears that a fix is in place in this commit: https://github.com/openwrt/openwrt/commit/1278d47beaabaa963b2956e81936269b7fea4003

As seen at: https://git.openwrt.org/?p=openwrt/openwrt.git;a=commitdiff;h=3bf602d740765db52d96733d172093a4b9a69915;hp=d4a21d75a9be5cf048bef126ea567c7ec61079a5

Which stems from: https://github.com/freifunk-gluon/gluon/pull/3084

**Pull request closed without merging**
